src/main/scala/firrtl/ir/IR.scala

```diff
@@ -211,8 +211,28 @@ abstract class Width extends FirrtlNode {
   }
 }
 /** Positive Integer Bit Width of a [[GroundType]] */
-case class IntWidth(width: BigInt) extends Width {
+object IntWidth {
+  private val maxCached = 1024
+  private val cache = new Array[IntWidth](maxCached)
+  def apply(width: BigInt): IntWidth = {
+    if (1 <= width && width <= maxCached) {
+      val i = width.toInt - 1
+      var w = cache(i)
+      if (w eq null) {
+        w = new IntWidth(width) ; cache(i) = w
+      }
+      w
+    } else new IntWidth(width)
+  }
+  // For pattern matching
+  def unapply(w: IntWidth): Option[BigInt] = Some(w.width)
+}
+class IntWidth(val width: BigInt) extends Width {
```

Comment:
For the purposes of the node counting function IntWidth needs to implement Product.
Also do we need to override hashcode? What is the default hashcode of a Scala class? We could easily just return the width.toInt